File: Leetcode/Graph/200_number-of-islands/main.py
from typing import List


class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        nr = len(grid)
        if nr == 0:
            return 0
        nc = len(grid[0])
        def dfs(r, c):
            grid[r][c] = '0'
            for x, y in [(r - 1, c), (r + 1, c), (r, c - 1), (r, c + 1)]:
                if 0 <= x < nr and 0 <= y < nc and grid[x][y] == '1':
                    dfs(x, y)
        sum = 0
        for i in range(nr):
            for j in range(nc):
                if grid[i][j] == "1":
                    dfs(i, j)
                    sum += 1
        return sum


if __name__ == '__main__':
    grid = [
        ["1", "1", "0", "0", "0"],
        ["1", "1", "0", "0", "0"],
        ["0", "0", "1", "0", "0"],
        ["0", "0", "0", "1", "1"]
    ]
    print(Solution().numIslands(grid))  # expect: 3

File: Leetcode/Graph/200_number-of-islands/main.cpp
//
// Created by bob on 2022/2/8.
//
#include<iostream>
#include<vector>

using namespace std;

class Solution {
private:
    /**
     * DFS算法，岛屿数量的子算法
     * @param grid 二维数组
     * @param r 遍历该顶点的横坐标
     * @param c 遍历该顶点的纵坐标
     */
    void dfs(vector<vector<char>> &grid, int r, int c) {
        int rLen = grid.size();
        int cLen = grid[0].size();
        grid[r][c] = '0';   // 将遍历过的值修改为0
        if (r - 1 >= 0 && grid[r - 1][c] == '1') dfs(grid, r - 1, c);
        if (r + 1 < rLen && grid[r + 1][c] == '1') dfs(grid, r + 1, c);
        if (c - 1 >= 0 && grid[r][c - 1] == '1') dfs(grid, r, c - 1);
        if (c + 1 < cLen && grid[r][c + 1] == '1') dfs(grid, r, c + 1);
    }

public:
    /**
     * 岛屿数量
     * @param grid 二维数组
     * @return 岛屿的数量
     */
    int numIslands(vector<vector<char>> &grid) {
        int rLen = grid.size();
        int cLen = grid[0].size();
        int sum = 0; 
        for (int r = 0; r < rLen; ++r) {
            for (int c = 0; c < cLen; ++c) {
                if (grid[r][c] == '1') {
                    dfs(grid, r, c);
                    sum += 1;
                }
            }
        }
        return sum;
    }
};

int main() {
    vector<vector<char>> grid = {
            {'1', '1', '0', '0', '0'},
            {'1', '1', '0', '0', '0'},
            {'0', '0', '1', '0', '0'},
            {'0', '0', '0', '1', '1'}
    };
    Solution s = Solution();
    int num = s.numIslands(grid);
    printf("%d", num);  // expect: 3
}